Hot fix

pollyannagogolook opened the pull request:
1. 在 callScreeningService ：
- 把 repository 改成 property inject ，而不是在 constructor
- 宣告一個允許所有來電的 response，並使用在 respondCall

2. 在 Activity:
- 新增 role 權限的要求

3. 在 repository:
- 新開一個 coroutine ，讓存 cache 的動作跟把資料回傳 viewModel 的動作同時發生
- 在只有一次性的要求中，使用 coroutine 而不是 flow
- 在 getInComingNumber 還是使用 flow ，原因是 viewModel 需要等這個值有值的時候才需要被賦值。而這個值被賦值的情境，是只有來電事件發生的時候。

pollyannagogolook committed:
Fix: should request Call Role Permission in Activity before callScreeningService activate

- Inject repository as CallScreeningService's property, rather than constructor.
- declare requestPermissionLauncher and requestRoleLauncher to record permssions.
- use checkPermissionAndRequestRole and requestCallRolePermission to ask user to give app permission.

pollyannagogolook committed:
Fix: Use Coroutine rather than Flow in RepositoryImpl

- Because the API is one-in-one-out, coroutine is more suitable than Flow, Thus change all functions's return type except getInComingNumber function.

- Use Flow in getInComingNumber function in Repository is for viewModel need to wait repository emit value to collect only when calling event happens.

- Use Coroutine for saving remoteData to Room for avoiding delaying response.

pollyannagogolook committed:
Fix: add respondCall in CallScreenService

- declare a response which allow all calls
- call respondToCall by the response and callData.

christclin commented on app/src/main/java/com/example/pollycall/call_detect/CallScreeningService.kt:
```diff
@@ -13,27 +13,46 @@ import javax.inject.Inject
  * Currently, all calls are allowed
  */
 @AndroidEntryPoint
 class CallScreeningService @Inject constructor(
```
沒有要 inject constructor 的話，這樣就可以了
```
class CallScreeningService : CallScreeningService()
```
然後一般建議幫自己的 extended class 取別的名字 e.g. PollyCallScreeningService

christclin commented on app/src/main/java/com/example/pollycall/data/PollyCallRepositoryImpl.kt:
```diff
@@ -34,81 +30,112 @@ class PollyCallRepositoryImpl @Inject constructor(
+
+    // get phone number from flow, pass to viewModel
+    override fun getInComingNumber(): Flow<String> = flow {
+        _inComingNumber?.let {
```
看起來 getIncomingNumber 和 _inComingNumber 都沒在使用了?

pollyannagogolook replied:
感謝大大眼尖！！這邊是我忘記改到了，另外剛剛在改的時候，有另外優化這段的邏輯：
1. 在 repository 設置 stateflow 負責裝 response，當敲 api 或 cache 時拿到值，就塞給這個 stateflow
2. 原本在 viewModel 設計 inComingNumber 的原因是，我原本先拿到號碼之後，會去敲一次 repository 的 searchSearchCall 。不過仔細想想我這樣會變成敲兩次這個 function (在 callScreeningService 上敲一次、在 viewModel 為了拿值又敲了一次)。覺得這樣不必要且耗時，所以我改成在 viewModel 去 collect 來自 repository 的 response。等到一拿到值，就會把文字產生並且送到 foreground service 裡面。

pollyannagogolook committed:
Fix: use getInComingNumber function to pass number to viewModel

- change getInComingNumber function's return type from flow to stateflow
- collect the function's value to trigger searchScreenCall function.

pollyannagogolook committed:
Refactor: viewModel get search response from repository directly

- delete the function to get InComingNumber in viewModel, but use stateflow to collect the searchResponse's changes.

- set stateflow to emit searchResponse in repository
